src/corelibs/U2Core/src/util/U2SequenceUtils.cpp

```diff
@@ -463,23 +463,19 @@ void U2SequenceImporter::startSequence(U2OpStatus& os,
 
 void U2SequenceImporter::addBlock(const char* data, qint64 len, U2OpStatus& os) {
     CHECK(len > 0, );
-    // derive common alphabet
-    const DNAAlphabet* blockAl = U2AlphabetUtils::findBestAlphabet(data, len);
-    CHECK_EXT(blockAl != nullptr, os.setError("Failed to match sequence alphabet!"), );
-
-    const DNAAlphabet* oldAl = U2AlphabetUtils::getById(sequence.alphabet);
-    const DNAAlphabet* resAl = blockAl;
-    if (oldAl != nullptr) {
-        if (oldAl->getType() == DNAAlphabet_AMINO && resAl->getType() == DNAAlphabet_NUCL) {
-            resAl = oldAl;
-        } else if (resAl->getType() == DNAAlphabet_AMINO && oldAl->getType() == DNAAlphabet_NUCL) {
-            oldAl = resAl;
-        } else {
-            resAl = U2AlphabetUtils::deriveCommonAlphabet(blockAl, oldAl);
+    for (int i = 0; i < len; i++) {
+        if (Q_UNLIKELY(!alphabetCharacterHit[data[i]])) {
+            alphabetCharacterHit[data[i]] = true;
```

Comment:
you do not need if() here. Just use `alphabetCharacterHit[data[i]] = true with no if.

Reply:
QBitArray already intialized with 'falses', so it is faster not to overwrite with the same value.

Reply:
Could you please share a performance test that shows that the optimization you do (and extra logic for it) has a positive effect? 

What I see here:
1) You read a memory  address
2) You have a branching instruction based on the value read.

In my suggested variant:
1) You read & write the same memory address
2) You do not have branching (if) - the loop contains less instructions and is easier to optimize by the compiler.

In case if you wrote the same value to memory that it stored before it may a no-op from the CPU memory management POV.